fiftyone/utils/iou.py

```diff
@@ -495,81 +495,141 @@ def _compute_bbox_ious(preds, gts, iscrowd=None, classwise=False):
+                        except Exception as e:
+                            inter = 0.0
+                            fou.handle_error(
+                                ValueError(
+                                    "Failed to compute intersection of predicted "
+                                    "object '%s' and ground truth object '%s'"
+                                    % (preds[i].id, gts[j].id)
+                                ),
+                                error_level,
+                                base_error=e,
+                            )
+
+                        if gt_crowd:
+                            union = pred_area
+                        else:
+                            union = pred_area + gt_area - inter
+
+                        iou = min(etan.safe_divide(inter, union), 1)
+
+                    ious[i, j] = iou
+    return ious
 
-                    iou = min(etan.safe_divide(inter, union), 1)
+def polyline2d_iou_matrix(preds, gts, iscrowd=None, classwise=False, gt_crowds=None):
+    num_pred = len(preds)
+    pred_labels = [pred.label for pred in preds]
+    # if is_symmetric:
+    #     num_gt = num_pred
+    #     gt_labels = pred_labels
+    # else:
+    num_gt = len(gts)
+    gt_labels = [gt.label for gt in gts]
+
+    if iscrowd is not None:
+        gt_crowds = [iscrowd(gt) for gt in gts]
+    elif gt_crowds is None:
+        gt_crowds = [False] * num_gt
+
+    ious = np.zeros((num_pred, num_gt))
+
+    for j, (gt, gt_label, gt_crowd) in enumerate(
+        zip(gts, gt_labels, gt_crowds)
+    ):
+        for i, (pred, pred_label) in enumerate(
+            zip(preds, pred_labels)
+        ):
+            # if is_symmetric and i < j:
+            #     iou = ious[j, i]
+            # elif is_symmetric and i == j:
+            #     iou = 1
+            if classwise and pred_label != gt_label:
+                continue
+            ious[i, j] = _compute_object_polyline_similarity(gt, pred)
+    return ious
+
+def _compute_object_polyline_similarity(gt, pred):
+    gtp = np.array(gt.points, dtype=float)[0]
+    predp = np.array(pred.points, dtype=float)[0]
+    each_dist =[]
+    # Use extent of GT points as proxy for box area
+    scale = np.sqrt(np.prod(np.nanmax(gtp, axis=0) - np.nanmin(gtp, axis=0)))
+    scale = np.maximum(0.0, np.minimum(scale, 1.0))
+    for ind in np.arange(len(gtp)):
 
-                ious[i, j] = iou
+        if not np.isfinite(gtp[ind]).all():
+            continue
+        if np.isfinite(predp[ind]).all():
+            dists = np.linalg.norm(gtp[ind]-predp[ind])
+        else:
+            dists = 1
+        each_dist.append(np.exp(-(dists**2) / (2 * (scale**2))))
+    return np.mean(each_dist)
+    
+def _compute_polyline_ious(
+    preds, gts, error_level, iscrowd=None, classwise=False, gt_crowds=None
+):
+    is_symmetric = preds is gts
 
-        return ious
+    are_all_closed = np.sum([p ["closed"] for p in preds])
```

Comment:
Whether a `Polyline` object represents a polyline or a polygon is slightly undefined in our current data model. There are two parameters:

```
        closed (False): whether the shapes are closed, i.e., and edge should
            be drawn from the last vertex to the first vertex of each shape
        filled (False): whether the polyline represents polygons, i.e., shapes
            that should be filled when rendering them
```

But this leaves open the possibility that the user set `closed=False` and instead duplicated the first node at the end of the `points` list to manually "close" the shape.

Arguably checking for `filled=True` is a more suitable check for polygons, since if the user provides `filled=False` the shape will be rendered as lines rather than filled polygons in the App. However, since we have not been strict about what determines a polygon vs a polyline, potentially a user could have a closed shape that they intend to represent a polygon but for whatever reason they set `filled=False` because they prefer for it not to be filled in when visualizing in the App.

Reply:
Actually I'm now looking at the documentation of the main `compute_ious()` method on line 40 and I see:

```
    For polylines, IoUs are computed assuming the shapes are solid (filled),
    regardless of their ``filled`` attributes.
```

so I recommend using `filled=True/False` to distinguish between polygons and polylines, and correspondingly update the documentation on line 40 to read:

```
    For :class:`fiftyone.core.labels.Polyline` objects, IoUs are computed
    as solid shapes when ``filled=True` and are computed using
    `object keypoint similarity <https://cocodataset.org/#keypoints-eval>`_
    when ``filled=False``.
```

Reply:
@brimoor  fixed in commit https://github.com/voxel51/fiftyone/pull/3930/commits/17665146f562bd8e9936d700c7cdc7be06a3257e